src/pset/pset.3
.\"(c) Copyright 1992, 1993 by Panagiotis Tsirigotis
.\"All rights reserved.  The file named COPYRIGHT specifies the terms 
.\"and conditions for redistribution.
.\"
.\" $Id: pset.3,v 1.1 2001/05/26 22:04:50 seth Exp $
.TH PSET 3X "23 April 1993"
.SH NAME
pset_create, pset_destroy, pset_add, pset_insert, pset_remove, pset_delete, pset_remove_index, pset_clear, pset_count, pset_pointer, pset_compact, pset_sort, pset_apply - routines that handle pointer sets
.SH SYNOPSIS
.LP
.nf
.ft B
#include "pset.h"
.LP
.ft B
pset_h pset_create( alloc_start, alloc_step )
unsigned alloc_start, alloc_step ;
.LP
.ft B
void pset_destroy( pset )
pset_h pset ;
.LP
.ft B
ANY_TYPE *pset_add( pset, ptr )
pset_h pset ;
ANY_TYPE *ptr ;
.LP
.ft B
void *pset_insert( pset, ptr )
pset_h pset ;
void *ptr ;
.LP
.ft B
void pset_remove( pset, ptr )
pset_h pset ;
ANY_TYPE *ptr ;
.LP
.ft B
void pset_delete( pset, ptr )
pset_h pset ;
void *ptr ;
.LP
.ft B
void pset_remove_index( pset, index )
pset_h pset ;
unsigned index ;
.LP
.ft B
void pset_clear( pset )
pset_h pset ;
.LP
.ft B
unsigned pset_count( pset )
pset_h pset ;
.LP
.ft B
void *pset_pointer( pset, index )
pset_h pset ;
unsigned index ;
.LP
.ft B
void pset_compact( pset )
pset_h pset ;
.LP
.ft B
void pset_sort( pset, compfunc )
pset_h pset ;
int (*compfunc)() ;
.LP
.ft B
void pset_apply( pset, func, arg )
pset_h pset ;
void (*func)() ;
void *arg ;
.SH DESCRIPTION
This library provides functions that handle sets of pointers. Pointers
can be inserted and deleted from sets and the sets can be enumerated.
Pointers are inserted in sets in no particular order. However it is 
guaranteed
that a sequence of insertions will result in a set which if enumerated
will provide the pointers in the same order in which they were inserted
(assuming no intervening deletions).
.LP
.B pset_create()
creates a pointer set.
It will allocate space for
.I alloc_start
pointers before returning.
.I alloc_step
determines the amount by which the set size is increased in case of
overflow. If any of these parameters is 0, a default value is used.
.LP
.B pset_destroy()
destroys the specified pointer set.
.LP
.B pset_add()
is a macro that adds a pointer to the specified set.
The pointer can be of any type.
.LP
.B pset_insert()
inserts a pointer to the specified set.
This is the same operation as
.B pset_add().
.LP
.B pset_remove()
removes a pointer from the specified set.
.LP
.B pset_delete()
deletes a pointer from the specified set.
This is the same operation as
.B pset_remove().
.LP
.B pset_remove_index()
removes the pointer that is at position
.I index
in the set.
.I index
should be in the range [0, \fBpset_count(pset)\fP) (but there is no
check to enforce this).
After this operation, the
.I index
position will be occupied by another pointer (unless it was the last one).
Note that this call is a macro, and it may evaluate the
.I index
argument more than once.
.LP
.B pset_clear()
removes all pointers from the specified set.
.LP
.B pset_count()
returns the number of pointers in the specified set.
.LP
.B pset_pointer()
returns the pointer at position
.I index
in the specified set.
.I index
must be between 0 and
.B "pset_count(pset)."
.B pset_pointer()
is a macro and it can also be used in the left-hand side of assignments.
.LP
.B pset_compact()
removes all NULL pointers from
.I pset.
.LP
.B pset_sort()
sorts the pointers in
.I pset
using the specified function.
.I compfunc
is invoked with 2 arguments that are pointers pointing to pointers stored in 
.I pset.
For example, if the pset holds pointers to objects of type T, then
the function F whose address is in
.I compfunc
should be defined as:
F( T **p1, T **p2 ).
.br
.I compfunc 
should return a negative, zero or positive value 
if its first argument is less than, equal to, or greater than its
second argument.
.LP
.B pset_apply()
applies
.I func
to all pointers in
.I pset.
If 
.I arg
is not
.SM NULL
the function is invoked as:
.RS
(*func)( arg, p )
.RE
where
.I p
is a pset pointer.  If 
.I arg
is
.SM NULL
the function is invoked as:
.RS
(*func)( p )
.RE
.SH EXAMPLE
The following code fragment reads lines from standard input
and places them in a pset. Then it sorts the pset, prints the
sorted contents to standard output and then it eliminates duplicate
lines (which it also prints to standard output).
.RS
.sp 1
.ft B
.nf
pset_h ph ;
char buf[ 80 ] ;
unsigned u ;
int compstr() ;
void printstr() ;
.sp 1
ph = pset_create( 0, 0 ) ;
while ( gets( buf ) )
.RS
pset_add( strcpy( malloc( strlen( buf ) + 1 ), buf ) ) ;
.RE
pset_sort( ph, compstr ) ;
for ( u = 0 ; u < pset_count( ph ) ; u++ )
.RS
printf( "%s\\n", (char *) pset_pointer( ph, u ) ) ;
.RE
.RE
.fi
.ft R
.LP
The function
.I compstr()
follows:
.sp 1
.RS
.ft B
.nf
int compstr( p1, p2 )
.RS
char **p1, **p2 ;
.RE
{
.RS
return( strcmp( *p1, *p2 ) ) ;
.RE
}
.RE
.SH "RETURN VALUES"
.LP
.I pset_h
is a pointer type. Functions that return
.I pset_h
will return
.SM NULL
to indicate an error.
.LP
.B pset_create()
returns a pointer set handle or
.SM NULL
if it fails.
.LP
.B pset_add()
returns its second argument if successful or
.SM NULL
if it fails.
.LP
.B pset_insert()
returns its second argument if successful or
.SM NULL
if it fails.
.LP
.B pset_count()
always returns the number of pointers in the set.
.LP
.B pset_pointer()
always returns a pointer. There is no check if the specified index is within
range.
.SH BUGS
.LP
.B pset_add(),
.B pset_remove(),
.B pset_remove_index(),
.B pset_count(),
.B pset_clear(),
.B pset_pointer()
and 
.B pset_sort()
are macros, therefore the \fI&\fR operator cannot be applied to them.


ChangeLog
2003-07-16  Seth Robertson  <seth@startide.sysdetect.com>

	* src/fsma/fsma.c (_fsm_free): Prevent infinite loops on fsm_free
	of corrupted memory

2003-07-09  James Tanis  <jtt@hawking.sysdetect.com>

	* src/dict/dll.c (dll_error_reason): Export pulibcally

2003-06-13  Alexander Dupuy  <dupuy@sysdetect.com>

	* include/dict.h, include/ht.h, include/pq.h: avoid stupid void * #1311

2003-06-07  Seth Robertson  <seth@startide.sysdetect.com>

	* src/dict/ht.c (ht_create): argsp may change size between
	external and internal users due to HASH_STATS define.

2003-05-16  Seth Robertson  <seth@startide.sysdetect.com>

	* src/dict/ht.c (ht_destroy): Allow clean up of all memory when
	insure is in use and fsma is replaced by malloc

2003-05-10  Alexander Dupuy  <dupuy@sysdetect.com>

	* include/dict.h (DICT_STATS): keep performance stats for hash tables

	* src/dict/htimpl.h (struct ht_stats): hash table performance stats
	(ENTRY_IS_FULL): kludge to detect bucket collisions

	* src/dict/ht.c: instrumented to keep performance stats for hash tables
	(print_hashstats): print performance stats

2003-04-28  Alexander Dupuy  <dupuy@sysdetect.com>

	* src/dict/ht.c (ht_delete): microscopic performance tweak
	[Seth got the big win already]

2003-04-16  James Tanis  <jtt@saint.sysdetect.com>

	* src/fsma/fsma.c (fsm_threaded_makeready): Only allow users to
	actually alter the value of thread_overried if we are compiled
	with threads.

2003-04-08  Seth Robertson  <seth@startide.sysdetect.com>

	* src/fsma/fsma.c (fsm_threaded_makeready): Provide interface for
	people to specify this will be a threaded program and to enforce
	thread safe memory protection.

2003-04-03  Seth Robertson  <seth@startide.sysdetect.com>

	* src/pq/hpq.c: Make thread-safe, through PQ_THREADED_SAFE

2003-03-31  Seth Robertson  <seth@winwood.sysdetect.com>

	* src/dict/*.c src/fsma/*.c: Make thread-safe, through
	DICT_THREADED_SAFE and DICT_THREADED_MEMORY

2002-12-10  James Tanis  <jtt@stevedore.sysdetect.com>

	* src/pq/Makefile: move -lpq to -lpqueue so as not to conflict
	with postgres

2002-11-11  James Tanis  <jtt@stevedore.sysdetect.com>

	* src/fsma/fsma.c, include/dict.h: opx_rollup_3 merge

2002-11-09  James Tanis  <jtt@stevedore.sysdetect.com>

	* include/dict.h, src/fsma/fsma.c: trheads merge

2002-11-06  Seth Robertson  <seth@stevedore.sysdetect.com>

	* src/fsma/fsma.c: Compile problems when using dmalloc

2002-11-05  Seth Robertson  <seth@stevedore.sysdetect.com>

	* include/dict.h: Define symbol for NOCOALESCE which makes it
	clear the only reason we are using it is because we are too lazy
	to make CLC thread-safe

2002-09-09  James Tanis  <jtt@stevedore.sysdetect.com>

	* src/dict/ht.c: insight nits

2002-09-05  Seth Robertson  <seth@stevedore.sysdetect.com>

	* src/dict/dll.c, ChangeLog: Do not create dlls w/o UNORDERED and
	w/o OOCMP function--coredump otherwise sometimes

2002-09-05  Seth Robertson  <seth@winwood.sysdetect.com>

	* src/dict/dll.c (dll_create): Default order is ordered.

	* src/dict/dll.c (dll_create): Use bitwise and for bits, instead
	of addition.  Prevented an argument sanity check from occurring.

2002-08-29  Seth Robertson  <seth@startide.sysdetect.com>

	* include/fsma.h (struct __fsma_header): AHHHHH-HHHHAAAAAAAA!
	The reference counter was previously a short.  Worker, for reasons
	too complicated to get into, can have more than 2^16th lists.
	If we happen (not unlikely) to destroy down to 2^16 (e.g. zero)
	fsma thinks all allocators are free and promptly proceeds to return
	the memory to the system.  Memory was free'd but still in use.
	BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA.

	* src/fsma/fsma.c (fsm_destroy): Perform assertion to verify we
	actually found the fsma allocator in the coalesce list.  Don't
	bother to realloc the coalesce list size down--it won't do no good
	nohow.

2002-08-27  Seth Robertson  <seth@startide.sysdetect.com>

	* src/dict/ht.c (ht_minimum): Serious and quite possibly
	bug-injecting misuse of a hint (cur_min).  The insert and delete
	routines were setting this to a bucket chain entry, not a table
	entry bucket (except when the insert routine had to perform a
	bucket chain expansion--when cur_min might be set to some other
	random value).  This usually was OK since bucket chain numbers are
	very low (0-3 range), but if a valid entry was in the table at
	slot zero or one, then ht_minimum might never see that item.

2002-08-14  James Tanis  <jtt@saint.sysdetect.com>

	* src/dict/ht.c (ht_minimum): Use cached minimum index as search start.
	(ht_do_insert): Update minimum index if less the current minimum.
	(ht_delete): Search for new minimum index if the index of the
	deleted node equals the current minimum.
	In addition add and obj_cnt so we don't try to perform expensive
	operations on empty hashes.

2002-05-13  James Tanis  <jtt@saint.sysdetect.com>

	* src/fsma/Makefile (BK_LARGE_LIB): Shared libfsma no longer
	breaks with libtool (on both linux and bsd).

2002-05-10  Alexander Dupuy  <dupuy@sysdetect.com>

	* src/fsma/Makefile (BK_LARGE_LIB): back out shared libfsma;
	this breaks libtool in a strange way that requires investigation

2002-05-10  James Tanis  <jtt@winwood.sysdetect.com>

	* src/fsma/fsma.c: Formatting and simplify coalesce code.

	* src/fsma/Makefile (BK_LARGE_LIB): Make libfsma shared.

	* src/dict/bst.c (bst_redelete): Improve efficiency by avoiding
	unnecessary checks and recusive calls.

2002-02-22  Alexander Dupuy  <dupuy@sysdetect.com>

	* include/dict.h (DICT_NUKE): delete contents and dict itself
	(_DICT_NUKE_WORK_LOOP): common code for DICT_NUKE, DICT_NUKE_CONTENTS

	* src/dict/dictimpl.h (HANDLE_ERROR): Eliminate unused id argument
	[use BK_FUNCNAME  here instead if you need it for debugging]

2002-01-31  Alexander Dupuy  <dupuy@sysdetect.com>

	* src/dict/Makefile (BK_LARGE_LIB): use ${LIBEXT} for libtool build

2002-01-09  Alexander Dupuy  <dupuy@sysdetect.com>

	* src/fsma/fsma.c: INSIGHT->INSURE


src/pset/pset.c
/*
 * (c) Copyright 1992, 1993 by Panagiotis Tsirigotis
 * All rights reserved.  The file named COPYRIGHT specifies the terms 
 * and conditions for redistribution.
 */

static char RCSid[] = "$Id: pset.c,v 1.1 2001/05/26 22:04:50 seth Exp $" ;
static char version[] = VERSION ;

#include "pset.h"
#include <stdlib.h>
#include "clchack.h"

#ifndef NULL
#define NULL						0
#endif

#define ALLOC_START				20
#define ALLOC_STEP				10

#define POINTER					__pset_pointer



/*
 * Create a pointer set and return a handle to it.
 * Some space is initially allocated for the set.
 */
pset_h pset_create(unsigned int alloc_start, unsigned int alloc_step)
{
	pset_h pset ;
	unsigned start ;

	pset = (pset_h) malloc( sizeof( struct __pset ) ) ;
	if ( pset == NULL )
		return( NULL ) ;
	
	start = ( alloc_start == 0 ) ? ALLOC_START : alloc_start ;
	pset->ptrs = (POINTER *) malloc( start * sizeof( POINTER ) ) ;
	if ( pset->ptrs == NULL )
	{
		free( (char *) pset ) ;
		return( NULL ) ;
	}

	pset->max = start ;
	pset->count = 0 ;
	pset->alloc_step = ( alloc_step == 0 ) ? ALLOC_STEP : alloc_step ;
	return( pset ) ;
}


/*
 * Destroy a pset
 */
void pset_destroy(pset_h pset)
{
	free( (char *) pset->ptrs ) ;
	free( (char *) pset ) ;
}


/*
 * Insert a pointer to a pset
 */
POINTER pset_insert(pset_h pset, __pset_pointer p)
{
	if ( pset->count >= pset->max )
	{
		unsigned new_max = pset->max + pset->alloc_step ;
		POINTER *new_ptrs ;

		new_ptrs = (POINTER *) realloc(
								(char *)pset->ptrs, new_max * sizeof( POINTER ) ) ;
		if ( new_ptrs == NULL )
			return( NULL ) ;
		pset->max = new_max ;
		pset->ptrs = new_ptrs ;
	}
	return( pset->ptrs[ pset->count++ ] = p ) ;
}


/*
 * Remove a pointer from a pset
 */
void pset_delete(register pset_h pset, register __pset_pointer ptr)
{
	register unsigned u = pset->count ;

	if ( u == 0 )
		return ;
	
	do
	{
		u-- ;
		if ( pset->ptrs[ u ] == ptr )
		{
			pset->ptrs[ u ] = pset->ptrs[ --pset->count ] ;
			return ;
		}
	}
	while ( u ) ;
}


/*
 * Create a pset iterator
 */
psi_h psi_create(pset_h pset)
{
	psi_h iter = (psi_h) malloc( sizeof( struct __pset_iterator ) ) ;

	if ( iter == NULL )
		return( NULL ) ;
	iter->pset = pset ;
	return( iter ) ;
}


